### PR TITLE
Enable to pass a bval Validator to BeanValidationProvider (it is more…

### PR DESCRIPTION
**Pull request by rmannibucau**
… than enough and faster than creating validators again and again) + enable to cache the fact no validation is needed

### CODE REVIEW

File: core/src/main/java/org/apache/cxf/validation/BeanValidationProvider.java [hunk 1]
```diff
@@ -111,46 +142,81 @@ private static void initFactoryConfig(Configuration<?> factoryCfg, ValidationCon
...
+                throw new ResponseConstraintViolationException(violations);
+            }
         }
     }
 
     public< T > void validateReturnValue(final T bean) {
-        final Set<ConstraintViolation< T > > violations = doValidateBean(bean);
+        Validator validator = factory.get();
+        if (runtimeCache != null && bean != null
+                && !runtimeCache.shouldValidateBean(validator, bean.getClass())) {
```
Comment (reta) [inline core/src/main/java/org/apache/cxf/validation/BeanValidationProvider.java:171]: Shouldn't it be `!runtimeCache.shouldValidateReturnedValue`?

File: core/src/main/java/org/apache/cxf/validation/BeanValidationProvider.java [hunk 2]
```diff
@@ -111,46 +142,81 @@ private static void initFactoryConfig(Configuration<?> factoryCfg, ValidationCon
...
+        final Set<ConstraintViolation< T > > violations = doValidateBean(factory.get(), bean);
         if (!violations.isEmpty()) {
             throw new ConstraintViolationException(violations);
         }
     }
 
-    private< T > Set<ConstraintViolation< T > > doValidateBean(final T bean) {
-        return factory.getValidator().validate(bean);
+    private< T > Set<ConstraintViolation< T > > doValidateBean(final Validator validator, final T bean) {
+        if (validator.getConstraintsForClass(bean.getClass()).isBeanConstrained()) {
```
Comment (reta) [inline core/src/main/java/org/apache/cxf/validation/BeanValidationProvider.java:188]: Probably call to `runtime.shouldValidateBean` is needed

File: core/src/main/java/org/apache/cxf/validation/BeanValidationProvider.java (hunk 1 above)
Comment (reta) [reply core/src/main/java/org/apache/cxf/validation/BeanValidationProvider.java:171]: Oh nevermind, there is no method in this call

File: core/src/main/java/org/apache/cxf/validation/BeanValidationProvider.java [hunk 3]
```diff
@@ -111,46 +142,81 @@ private static void initFactoryConfig(Configuration<?> factoryCfg, ValidationCon
...
 
-        return factory.getValidator().forExecutables();
+        public boolean shouldValidateReturnedValue(final Validator validator, final Method method) {
+            return returnedValues.computeIfAbsent(method, m -> validator.getConstraintsForClass(m.getDeclaringClass())
+                    .getConstraintsForMethod(m.getName(), method.getParameterTypes())
+                    .hasConstrainedReturnValue());
+        }
+
+        public boolean shouldValidateBean(final Validator validator, final Class<?> clazz) {
+            return types.computeIfAbsent(clazz, it -> validator.getConstraintsForClass(it).hasConstraints());
```
Comment (reta) [inline core/src/main/java/org/apache/cxf/validation/BeanValidationProvider.java:219]: Should you also check for `isBeanConstrained` and cache that?

File: core/src/main/java/org/apache/cxf/validation/BeanValidationProvider.java (hunk 2 above)
Comment (rmannibucau) [reply core/src/main/java/org/apache/cxf/validation/BeanValidationProvider.java:188]: think this call is always protected by a public call before which already have the exit condition - at least was the intent

**Comment by rmannibucau** [comment]
@reta thanks for the review, fixed the last comment, think others are ok (already "protected" by the caller)

**Review by reta — APPROVED** [review]

**Comment by reta** [comment]
@rmannibucau anything outstanding left for this pr? or good to merge?

**Comment by rmannibucau** [comment]
@reta good to merge for me

**Pull request merged**